### PR TITLE
Refactor sockets

### LINKED ISSUE #41: mirrord-layer deadlocks if accept is called when there's no data
### Bug Description

When accept is called before a connection is read from the agent, it locks the connections struct and there's a deadlock.

### Steps to Reproduce

Simulate a delay so that accept is called before a connection is read from the agent.

### Backtrace

_No response_

### Your operating system and version

MacOS

### Your Node version (`node --version`)

N/A

### Additional Info

_No response_

### LINKED ISSUE #42: When running with VSCode's Node debugger, mirrord hooks and breaks an internal socket
### Bug Description

VSCode's Node debugger uses a socket for some internal mechanism. Mirrord hooks it, breaking the flow. We should be able to overlook these sockets.

### Steps to Reproduce

Debug a node process in VSCode with mirrord-layer in LD_PRELOAD.

### Backtrace

_No response_

### Your operating system and version

MacOS

### Your Node version (`node --version`)

v17.6.0

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by aviramha**
Refactor sockets - beforehand we created socketpair on each socket creation, now we only do it when necessary using delayed execution of bind,listen.

closes #41 closes #42

### CODE REVIEW

File: mirrord-layer/src/sockets.rs [hunk 1]
```diff
@@ -1,240 +1,368 @@
...
 use queues::{IsQueue, Queue};
 use socketpair::{socketpair_stream, SocketpairStream};
 use tracing::{debug, error};
 
-pub struct Socket {
-    pub read_fd: SockFd,
-    pub read_socket: SocketpairStream,
-    pub write_socket: SocketpairStream,
-}
+use crate::{
```
Comment (eyalb181) [inline mirrord-layer/src/sockets.rs:19]: Why do you need use crate here, can't you just do 'mod'?

File: mirrord-layer/src/sockets.rs [hunk 2]
```diff
@@ -1,240 +1,368 @@
...
-                    write_socket: socket.write_socket,
-                    address,
-                    state: ConnectionState::Bound,
-                },
-            );
-        } else {
-            error!("No socket found for fd: {}", sockfd);
-        }
-    }
+impl Eq for Socket {}
```
Comment (eyalb181) [inline mirrord-layer/src/sockets.rs:60]: Why is this necessary?

File: mirrord-layer/src/sockets.rs [hunk 3]
```diff
@@ -1,240 +1,368 @@
...
+#[derive(Default)]
+struct State {
+    sockets: HashSet<Socket>,
+}
 
-    pub fn get_data_socket_address(&self, sockfd: SockFd) -> Option<SocketAddr> {
-        let sockets = self.data.lock().unwrap();
-        sockets.get(&sockfd).map(|socket| socket.address)
+lazy_static! {
+    static ref STATE: Mutex<State> = Mutex::new(State::default());
```
Comment (eyalb181) [inline mirrord-layer/src/sockets.rs:80]: Why not just keep a direct reference to the sockets?

File: mirrord-layer/src/sockets.rs [hunk 4]
```diff
@@ -1,240 +1,368 @@
...
+}
+
+#[inline]
```
Comment (eyalb181) [inline mirrord-layer/src/sockets.rs:83]: Why'd you decide to inline it?

File: mirrord-layer/src/sockets.rs [hunk 5]
```diff
@@ -1,240 +1,368 @@
...
+fn is_ignored_port(port: Port) -> bool {
+    port == 0 || (port > 50000 && port < 60000)
+}
+
+fn socket(domain: c_int, type_: c_int, protocol: c_int) -> RawFd {
```
Comment (eyalb181) [inline mirrord-layer/src/sockets.rs:88]: Why the underscore suffix for type_?

File: mirrord-layer/src/sockets.rs [hunk 6]
```diff
@@ -1,240 +1,368 @@
...
+            }
+            Some(socket) => socket,
+            None => {
+                debug!("bind: no socket found for fd: {}", &sockfd);
+                return unsafe { libc::bind(sockfd, addr, addrlen) };
+            }
+        }
+    };
+
+    let raw_addr = unsafe { OsSocketAddr::from_raw_parts(addr as *const u8, addrlen as usize) };
```
Comment (eyalb181) [inline mirrord-layer/src/sockets.rs:133]: Why not move this to the beginning of the function? Then you could edit the socket as you find it with get_mut and not have to lock the State twice

File: mirrord-layer/src/sockets.rs [hunk 7]
```diff
@@ -1,240 +1,361 @@
...
+lazy_static! {
+    static ref STATE: Mutex<State> = Mutex::new(State::default());
+}
+
+#[inline]
+fn is_ignored_port(port: Port) -> bool {
+    port == 0 || (port > 50000 && port < 60000)
+}
+
+fn socket(domain: c_int, type_: c_int, protocol: c_int) -> RawFd {
```
Comment (eyalb181) [inline mirrord-layer/src/sockets.rs:88]: I'm going to add my interpretation of what the hooks do, let me know if I didn't get it right, and if I did please add it as function documentation.
For socket:
Create the socket, add it to State, and return it to the calling application.

File: mirrord-layer/src/sockets.rs [hunk 8]
```diff
@@ -1,240 +1,361 @@
...
-            let mut buffer = [0; 1];
-            socket.read_socket.read_exact(&mut buffer).unwrap();
-            sockets.insert(sockfd, socket.address.port(), socket);
-            let mut queues = self.connection_queues.lock().unwrap();
-            let queue = queues.get_mut(&sockfd).unwrap();
-            queue.remove().unwrap()
-        } else {
-            error!("No socket found for fd: {}", sockfd);
-            0
+fn bind(sockfd: c_int, addr: *const sockaddr, addrlen: socklen_t) -> c_int {
```
Comment (eyalb181) [inline mirrord-layer/src/sockets.rs:116]: If it's an unignored port, map the address the application bound to the socket.

File: mirrord-layer/src/sockets.rs [hunk 9]
```diff
@@ -1,240 +1,361 @@
...
+            let ret = unsafe { libc::bind(sockfd, os_addr.as_ptr(), os_addr.len()) };
+            if ret != 0 {
+                error!(
+                    "listen: failed to bind socket ret: {:?}, addr: {:?}, sockfd: {:?}, errno: {:?}",
+                    ret, os_addr, sockfd, errno()
+                );
+                return ret;
+            }
+            let mut addr_len = os_addr.len();
+            let ret = unsafe { libc::getsockname(sockfd, os_addr.as_mut_ptr(), &mut addr_len) };
```
Comment (eyalb181) [inline mirrord-layer/src/sockets.rs:203]: Why is this necessary?

File: mirrord-layer/src/sockets.rs [hunk 10]
```diff
@@ -1,240 +1,373 @@
...
+                    OsSocketAddr::from(SocketAddr::new(IpAddr::V6(Ipv6Addr::UNSPECIFIED), 0))
+                }
+                _ => {
+                    // shouldn't happen
+                    debug!("unsupported domain");
+                    return libc::EINVAL;
+                }
+            };
+
+            let ret = unsafe { libc::bind(sockfd, os_addr.as_ptr(), os_addr.len()) };
```
Comment (eyalb181) [inline mirrord-layer/src/sockets.rs:194]: So does this call the hook, not find the socket, then call the actual function?

File: mirrord-layer/src/lib.rs [hunk 11]
```diff
@@ -1,151 +1,121 @@
...
-unsafe extern "C" fn accept4_detour(
-    sockfd: i32,
-    addr: *mut sockaddr,
-    addrlen: *mut socklen_t,
-    _flags: i32,
-) -> i32 {
-    accept_detour(sockfd, addr, addrlen)
+#[ctor]
+fn init() {
+    INIT.call_once(|| {
```
[the remark this answers is not in the document]
Comment (aviramha) [reply mirrord-layer/src/lib.rs:93]: erm, guess not

File: mirrord-layer/src/sockets.rs (hunk 2 above)
Comment (aviramha) [reply mirrord-layer/src/sockets.rs:60]: that's how you impl Eq which is required for HashSet.

File: mirrord-layer/src/sockets.rs (hunk 3 above)
Comment (aviramha) [reply mirrord-layer/src/sockets.rs:80]: I'm preparing for a more elaborate usecase, which might not exist! :)

Comment (aviramha) [reply mirrord-layer/src/sockets.rs:80]: (fixing)

File: mirrord-layer/src/sockets.rs (hunk 4 above)
Comment (aviramha) [reply mirrord-layer/src/sockets.rs:83]: because the over head of calling the function is stupid while it's just a simple condition wrapped in a function for beauty reasons

File: mirrord-layer/src/sockets.rs (hunk 6 above)
Comment (aviramha) [reply mirrord-layer/src/sockets.rs:133]: Because we want to remove it anway (if ignore we want to remove it,?

File: mirrord-layer/src/sockets.rs (hunk 9 above)
Comment (aviramha) [reply mirrord-layer/src/sockets.rs:203]: If you look at above lines, we provide port:0, this sets an unspecified port and this is how we retrieve what port it is.

File: mirrord-layer/src/sockets.rs (hunk 10 above)
Comment (aviramha) [reply mirrord-layer/src/sockets.rs:194]: yup

File: mirrord-layer/src/sockets.rs [hunk 12]
```diff
@@ -1,240 +1,506 @@
...
+
+unsafe extern "C" fn bind_detour(
+    sockfd: c_int,
+    addr: *const sockaddr,
+    addrlen: socklen_t,
+) -> c_int {
+    bind(sockfd, addr, addrlen)
+}
+
+fn listen(sockfd: RawFd, _backlog: c_int) -> c_int {
```
Comment (eyalb181) [inline mirrord-layer/src/sockets.rs:204]: Bind the socket to a fake, local port, and subscribe to the agent on the real port. Messages received from the agent on the real port will later be routed to the fake local port.

File: mirrord-layer/src/sockets.rs [hunk 13]
```diff
@@ -1,240 +1,506 @@
...
+            let ret = unsafe { libc::bind(sockfd, os_addr.as_ptr(), os_addr.len()) };
+            if ret != 0 {
+                error!(
+                    "listen: failed to bind socket ret: {:?}, addr: {:?}, sockfd: {:?}, errno: {:?}",
+                    ret, os_addr, sockfd, errno()
+                );
+                return ret;
+            }
+            let mut addr_len = os_addr.len();
+            let ret = unsafe { libc::getsockname(sockfd, os_addr.as_mut_ptr(), &mut addr_len) };
```
Comment (eyalb181) [inline mirrord-layer/src/sockets.rs:243]: // Resolve the fake local port that the socket was bound to

File: mirrord-layer/src/sockets.rs [hunk 14]
```diff
@@ -1,240 +1,506 @@
...
-            address,
-        };
-        self.data
-            .lock()
-            .unwrap()
-            .insert(read_fd, connection_id, data_socket);
-        read_fd
+    };
+
+    // We don't handle this socket, so restore state if there was any. (delay execute bind)
```
Comment (eyalb181) [inline mirrord-layer/src/sockets.rs:311]: What do you mean here with "we don't handle this socket"? Because it's for outgoing communication?

Comment (aviramha) [reply mirrord-layer/src/sockets.rs:321]: yes

File: mirrord-layer/src/sockets.rs [hunk 15]
```diff
@@ -1,240 +1,506 @@
...
+    listen(sockfd, backlog)
+}
+
+fn connect(sockfd: RawFd, address: *const sockaddr, len: socklen_t) -> c_int {
+    debug!("connect called");
 
-        if let Some(data) = self.pending_data.lock().unwrap().remove(&connection_id) {
-            debug!("writing pending data for connection_id: {}", connection_id);
-            write_socket.write_all(&data).unwrap();
+    let socket = {
```
[the remark this answers is not in the document]
Comment (eyalb181) [reply mirrord-layer/src/sockets.rs:312]: ?

File: mirrord-layer/src/sockets.rs (hunk 13 above)
Comment (aviramha) [reply mirrord-layer/src/sockets.rs:254]: I added it in the function doc?

**Review by eyalb181 — APPROVED** [review]
